src/arcsf/data/data_module.py

```diff
@@ -0,0 +1,207 @@
+        forgotten_fact_fraction,
+        random_seed=random_seed,
+    )
+
+    return data
+
+
+def QAformatter_basic(QA: tuple[str]) -> str:
+    """
+    Basic QA formatter which accepts a tuple outputs:
+
+    "Question: [input question]\nAnswer: [input answer]"
+
+    args:
+        - QA: Tuple of question answer pair
+    returns:
+        - full_text: formatted question--answer pair
+    """
+    Question, Answer = QA
+    question_formatting = "Question: " + Question
+    answer_formatting = "\nAnswer: " + Answer
+    full_text = question_formatting + answer_formatting
+    return full_text
+
+
+class EvalQADataSet(Dataset):
+    """
+    Question answer format dataset, __getitem__ returns a tokenized question--answer
+    pair as a tuple. There is an option to output the answers using "I don't know"
+    synonyms by specifying loss_type as "idk".
+    """
+
+    def __init__(
+        self,
+        data,
+        tokenizer,
+        qa_formatter,
+        split,
+        loss_type,
+    ):
+        super().__init__()
+        self.tokenizer = tokenizer
+        self.qa_formatter = qa_formatter
+        self.loss_type = loss_type
+        forget_data, retain_data = data
+
+        split_dict = {"retain": retain_data, "forget": forget_data}
+        self.data = split_dict[split]
+
+        if loss_type == "idk":
+            with open("src/arcsf/data/idk.jsonl") as idk_file:
+                self.idk = idk_file.read().splitlines()
+            self.answer_sampler = self.idk_sampler
+        else:
+            self.answer_sampler = self.forget_sampler
+
+    def answer_sampler():
+        return None
+
+    def idk_sampler(self, _):
+        rand_pos = torch.randint(0, len(self.idk), (1,)).item()
+        return self.idk[rand_pos]
+
+    def forget_sampler(self, forget_row):
+        return forget_row["answer"]
+
+    def __len__(self):
+        return len(self.data)
+
+    def __getitem__(self, idx):
+        input = self.data[idx]["question"]
+
+        target = self.answer_sampler(self.data[idx]["answer"])
+
+        return self.tokenizer(input), self.tokenizer(target)
+
+
+class FinetuneDataset(Dataset):
+    """
+    Finetune version of the dataset, __getitem__ returns a sample taken either from
+    retain, forget subsets, or a combination of both. Samples are formatted using a
+    question formatter allowing for autoregression. There is an option to output
+    samples using "I don't know" synonyms by specifying loss_type as "idk".
+    """
+
+    def __init__(
+        self,
+        data,
+        tokenizer,
+        qa_formatter,
+        split,
+    ):
+        super().__init__()
+        self.tokenizer = tokenizer
+        self.qa_formatter = qa_formatter
+
+        forget_data, retain_data = data
+
+        split_dict = {
+            "retain": retain_data,
+            "forget": forget_data,
+            "all": ConcatDataset([retain_data, forget_data]),
+        }
+        self.data = split_dict[split]
```

Comment:
`FinetuneDataset` and `EvalQADataset` both only use a single data split (at least at the moment), would it be easier if they were created only with the data they need (so the data argument was a single dataset not a tuple of forget and retain)? That would then also avoid needing this dict/concat step here, which feels a bit odd (start with whole dataset, split it, combine it again).

Reply:
Ok, will investigate this

Reply:
Addressed this now here and in tests, though `load_tofu` outputs a tuple, so its not necessarily obvious which split we are passing to the class using the index. Also these will now operate differently to the `ForgetDataset`. But I suppose this can be addressed down the line.